[JuhromeAlexander] iP

JuhromeAlexander opened the pull request:
# Donezo
> "The secret of getting ahead is getting started." - Mark Twain [(Source)](https://www.reddit.com/r/quotes/comments/1181xaq/the_secret_of_getting_ahead_is_getting_started/?rdt=59982)

This Donezo Bot is your all in one bot to help you get ahead in life! It's:
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is:
1. Download this from [here.](https://github.com/JuhromeAlexander/ip/releases/tag/A-Jar)
2. Open up your terminal and head to where you downloaded it
3. Run the following command:
```bash
java -jar "Donezo.jar"
```
4. Explore the application! 😄 

And it is **FREE**!

Features:
- [x] Managing Tasks
- [x] Managing Deadlines
- [x] Managing Events

If you are a Java Programmer, you can use it to practice Java too. Here's the ```main``` method:
```java
public static void main(String[] args) {
        Donezo donezo = new Donezo();
        try {
            donezo.run();
        } catch (DonezoException e) {
            System.out.println(e);
        }
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

JuhromeAlexander committed:
refactor: rename chatbot

Renamed chatbot from Duke to Donezo. Replaced all traces of Duke with
Donezo

JuhromeAlexander committed:
feat: add echo and exit functionality to chatbot

JuhromeAlexander committed:
feat: add list and basic storage functionality to chatbot

Add the ability to store user input and display them back to the user
when requested

JuhromeAlexander committed:
feat: Add ability to mark task as complete and incomplete

JuhromeAlexander committed:
feat: add support for tracking 3 new types of tasks

Added support for the Chatbot to track ToDos, Events and Deadlines
tasks.

JuhromeAlexander committed:
fix: fix deadline description argument

Fixed bug where Deadline Task will be created with 'deadline' as its
description instead of the actual description of the task

JuhromeAlexander committed:
feat: add error handling for the 3 types of task

Added proper handling to deal with incorrect inputs from the users of
the chatbot

JuhromeAlexander committed:
feat: add support for deleting tasks from list

Refactored the logic so that the delete functionality is in the same
method as the other commands, for easier error handling

JuhromeAlexander committed:
test: added test for main functionality of chatbot

Testing the list, event, todo, deadline, delete commands functionality

JuhromeAlexander committed:
test: add test support for mark/unmark commands

JuhromeAlexander committed:
feat: add save functionality for deadline tasks

JuhromeAlexander committed:
chore: modify .gitignore to ignore files in data folder

JuhromeAlexander committed:
fix: fix unmark/mark and delete commands

When using unmark and delete, call to Storage.deleteFromFile function
fires, saving the information to data/tasks.txt

Storage.deleteFromFile does not delete the information from
data/tasks.txt properly, causing the program to crash in the next
instance of running

Let's,
* update the Storage.deleteFromFile method to properly delete information
from the data/tasks.txt
* fix the unmark/mark and delete commands to use the proper Storage and
Parser class methods

JuhromeAlexander committed:
feat: add save functionality for todo tasks

JuhromeAlexander committed:
feat: add save functionality for event tasks

JuhromeAlexander committed:
fix: fix behaviour when initial data files/folders do not exist

JuhromeAlexander committed:
Merge branch 'branch-Level-7'

JuhromeAlexander committed:
feat: add localdatetime support for event tasks

JuhromeAlexander committed:
feat: add localdatetime support for deadline tasks

JuhromeAlexander committed:
Merge branch 'branch-Level-8'

JuhromeAlexander committed:
refactor: extract out related code to different classes

JuhromeAlexander committed:
refactor: refactor list command into separate class

JuhromeAlexander committed:
refactor: refactor unmark/mark command into separate classes

JuhromeAlexander committed:
refactor: refactor delete command into separate class

JuhromeAlexander committed:
refactor: refactor deadline command to separate class

JuhromeAlexander committed:
refactor: refactor event command into separate class

JuhromeAlexander committed:
refactor: refactor todo command into separate file

JuhromeAlexander committed:
chore: remove unnecessary comments in main Donezo class

JuhromeAlexander committed:
Merge branch 'branch-A-MoreOOP'

JuhromeAlexander committed:
refactor: refactored classes into packages

JuhromeAlexander committed:
Merge branch 'add-gradle-support'

JuhromeAlexander committed:
test: add tests for ParserStorage and Storage Classes

JuhromeAlexander committed:
build: modify build.gradle to allow shadow plugin to work

JuhromeAlexander committed:
feat: add search functionality to chatbot

JuhromeAlexander committed:
docs: add javadoc comments to classes and methods

JuhromeAlexander committed:
style: remove unnecessary and unused lines

JuhromeAlexander committed:
Merge branch 'branch-A-CodingStandard'

JuhromeAlexander committed:
Merge branch 'branch-Level-9'

JuhromeAlexander committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/donezo/Storage.java
#	src/main/java/donezo/UI.java

JuhromeAlexander committed:
docs: add javadoc to FindCommand Class and related methods

JuhromeAlexander committed:
style: add checkstyle.xml and suppressions.xml

JuhromeAlexander committed:
style: fix checkstyle violations in the commands package

JuhromeAlexander committed:
style: fix checkstyle violations in the donezo package

JuhromeAlexander committed:
style: fix checkstyle violations in the exceptions package

JuhromeAlexander committed:
style: fix checkstyle violations in the parser package

JuhromeAlexander committed:
style: fix checkstyle violations in the tasks package

JuhromeAlexander committed:
style: fix checkstyle violations of unit tests

JuhromeAlexander committed:
Merge branch 'branch-A-Checkstyle'

shelialia commented on src/main/java/donezo/TaskList.java:
```diff
@@ -0,0 +1,68 @@
+ * It also allows for obtaining the total number of tasks in the list.
+ */
+public class TaskList {
+    private ArrayList<Task> taskList;
```
Since taskList is an ArrayList<Task>, it represents a collection of tasks. To align with coding standards, consider renaming taskList to tasks for better readability and consistency.

shelialia commented on src/main/java/donezo/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,52 @@
+import java.io.IOException;
+
+import donezo.TaskList;
+import donezo.exceptions.DonezoException;
```
Good that the order of your import statements is always consistent, starting from java imports, and then to your own file imports

shelialia commented on src/main/java/donezo/Donezo.java:
```diff
@@ -0,0 +1,149 @@
+        
+        while (!userInput.equals("bye")) {
+            String taskType = parser.parseCommand(userInput);
+            switch (taskType.toLowerCase()) {
```
Good that you did not make a mistake in terms of indentation for the switch-case statement 👍

shelialia commented on src/main/java/donezo/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,52 @@
+        String deadlineDescription =
+                userInput.substring(userInput.indexOf("deadline") + 9, userInput.indexOf("/by")).trim();
+        if (deadlineDescription.isBlank()) {
+            throw new DonezoException(
```
Good that you ensured that your line length is no more than 120 characters, and indented where necessary 👍

shelialia reviewed (commented):
LGTM. Good job on following the coding standard closely, and making clear and JavaDocs comments.

JuhromeAlexander committed:
refactor: refactor UI Class and subclasses to support GUI

JuhromeAlexander committed:
feat: add GUI support for Donezo bot

JuhromeAlexander committed:
fix: fix bug where greeting message did not show upon GUI launch

JuhromeAlexander committed:
docs: add javadoc comments for all new methods and classes

JuhromeAlexander committed:
Merge branch 'branch-Level-10'

JuhromeAlexander committed:
refactor: rename taskList field in TaskList.java to tasks

Currently, the field taskList in taskList.java is used to represent
an ArrayList of tasks

Rename the private field "taskList" to "tasks" throughout the codebase
to improve clarity and consistency, as well as to follow the
specified coding guidelines for the course

This refactoring does not change any functionality, only
the naming to better reflect the variable's purpose.

JuhromeAlexander committed:
test: add assert statements to ensure index is greater than 0

Current methods in ParserStorage do not have any checks to ensure
whether the index of what they are getting is correct.

Any erroneous input from the user may cause the task to not be added
properly

Add assert statements to ensure that all the indexes are more than 0

This ensures that the data is properly parsed before being used.

JuhromeAlexander committed:
refactor: add assertCheck method to command class and subclasses

Currently, command subclasses are not checking whether their
Storage, UI, userInput and taskList dependencies are being injected
properly before executing the commands in the method.

This needs to change as the user has the option of using two different
UI classes depending on what type of UI the user is using. If the
dependency is not injected properly, it will cause the application to
not work properly.

Add a new `testAssert(String userInput, TaskList taskList)` method to
the base command class and called it in all command subclasses.

This change improves the robustness of our commands without altering
their functional behavior.

JuhromeAlexander committed:
Merge pull request #2 from JuhromeAlexander/branch-A-Assertions

branch-A-Assertions

JuhromeAlexander committed:
Merge branch 'master' of https://github.com/JuhromeAlexander/ip

* 'master' of https://github.com/JuhromeAlexander/ip:
  refactor: add assertCheck method to command class and subclasses
  test: add assert statements to ensure index is greater than 0

JuhromeAlexander committed:
Merge branch 'master' of https://github.com/JuhromeAlexander/ip into branch-A-CodeQuality

* 'master' of https://github.com/JuhromeAlexander/ip:
  refactor: add assertCheck method to command class and subclasses
  test: add assert statements to ensure index is greater than 0

JuhromeAlexander committed:
Merge pull request #1 from JuhromeAlexander/branch-A-CodeQuality

refactor: rename taskList field in TaskList.java to tasks

JuhromeAlexander committed:
Merge branch 'master' of https://github.com/JuhromeAlexander/ip

* 'master' of https://github.com/JuhromeAlexander/ip:
  refactor: rename taskList field in TaskList.java to tasks

JuhromeAlexander committed:
refactor: rename class to TaskStorage to indicate type of storage

JuhromeAlexander committed:
refactor: refactor storage class to interface for better abstraction

JuhromeAlexander committed:
refactor: refactor tasklist to lists package

JuhromeAlexander committed:
refactor: create ItemList superclass for better abstraction

JuhromeAlexander committed:
fix: fix all errors related to Types due to ItemList implementation

JuhromeAlexander committed:
feat: add ability to create, store and list simple notes

JuhromeAlexander committed:
Merge branch 'branch-D-Notes'

JuhromeAlexander committed:
feat: add auto-resizing to elements in GUI mode

JuhromeAlexander committed:
docs: add Tasks Commands to user guide

JuhromeAlexander committed:
feat: add delete functionality for notes

JuhromeAlexander committed:
fix: fix array out of bounds error for mark/unmark commands

JuhromeAlexander committed:
docs: update README.md to contain all Donezo features

JuhromeAlexander committed:
fix: fix errors where given parameter values did not match

JuhromeAlexander committed:
Merge branch 'branch-A-MoreErrorHandling' into branch-A-UserGuide

JuhromeAlexander committed:
fix: fix date input validation behaviour to cover more ranges of inputs

JuhromeAlexander committed:
fix: fix bug where initial 4 characters of content not loading

JuhromeAlexander committed:
Merge branch 'branch-A-BetterGui' into branch-A-UserGuide

JuhromeAlexander committed:
Merge branch 'branch-A-UserGuide'

JuhromeAlexander committed:
refactor: add more personalised greeting message for GUI

JuhromeAlexander committed:
feat: add title to javafx window

JuhromeAlexander committed:
style: add newline for command to improve output readability

JuhromeAlexander committed:
docs: add Ui.png for IP requirement

JuhromeAlexander committed:
docs: clean up README.md for better readability

JuhromeAlexander committed:
docs: add newline breaks for better readability

JuhromeAlexander committed:
fix: fix error where wrong error message is printed when list is empty
